### PR TITLE
refcount refactor

### PR DESCRIPTION
**Pull request by sklam**
* More NRT debug utils
* More tests using the above
* Less auto insertion of refcount op and more explicit in internal implementation.

### CODE REVIEW

**Comment by sklam** [comment]
This fixes #1123 and removes the workaround in #1122

**Comment by sklam** [comment]
To make it easier for reviewers, I will highlight the major changes.

**Change 1: MemInfo refcount is now initialized to 1**

Affected:
* NRT nopython<--->cpython bridge
* Refct prune pass

**Change 2: internal implementation explicitly refcount semantic explicitly**

3 function added, starting at https://github.com/numba/numba/pull/1318/files#diff-91719f9201134d288646d22bb9e01ecaR318

They denote if the implementation is:
* returning a new reference: https://github.com/numba/numba/pull/1318/files#diff-91719f9201134d288646d22bb9e01ecaR318
* returning a borrow reference: https://github.com/numba/numba/pull/1318/files#diff-91719f9201134d288646d22bb9e01ecaR325.  So it needs to incref before returning to maintain the semantic that all function returns a new reference.
* returning a non-nrt object: https://github.com/numba/numba/pull/1318/files#diff-91719f9201134d288646d22bb9e01ecaR336

We can enforce that all internal implementation must return with one of the above function.  In fact, I did that during development by introducing a ``cgutils.NewRef`` and checking at ``lowering.py``.  The ``NewRef`` object is removed (https://github.com/sklam/numba/commit/7cc15e46a68339bfc1f118f062b51d636f153e44) after all the tests are passing.  Perhaps, we may want to have something similar to avoid making refcount mistake.

File: numba/generators.py [hunk 1]
```diff
@@ -211,24 +217,17 @@ def lower_finalize_func_body(self, builder, genptr):
         Lower the body of the generator's finalizer: decref all live
         state variables.
```
Comment (seibert) [inline numba/generators.py]: Is this docstring still accurate about decref'ing the state variables now that the second block of code has been removed?

Comment (sklam) [inline numba/generators.py]: Yes.  The removed block were double-freeing some states.

**Comment by seibert** [comment]
OK, this looks good to me.  If @pitrou is happy with this, I'd say we are ready to merge.

File: numba/lowering.py [hunk 2]
```diff
@@ -531,9 +539,12 @@ def stararg_handler(index, param, vars):
                 argvals = [the_self] + argvals
 
             res = impl(self.builder, argvals)
+            res = res
```
Comment (pitrou) [inline numba/lowering.py]: ?

File: numba/runtime/_nrt_python.c [hunk 3]
```diff
@@ -156,7 +168,7 @@ int MemInfo_init(MemInfoObject *self, PyObject *args, PyObject *kwds) {
     if(PyErr_Occurred()) return -1;
     self->meminfo = (MemInfo*)raw_ptr;
     self->defer = 0;
-    NRT_MemInfo_acquire(self->meminfo);
+    assert (NRT_MemInfo_refcount(self->meminfo) > 0 && "0 refcount");
```
Comment (pitrou) [inline numba/runtime/_nrt_python.c]: It doesn't sound very valuable to have asserts if nobody compiles in debug mode :-)

File: numba/targets/arrayobj.py [hunk 4]
```diff
@@ -188,7 +187,9 @@ def getiter_array(context, builder, sig, args):
     iterobj.index = indexptr
     iterobj.array = array
 
-    return iterobj._getvalue()
+    res = iterobj._getvalue()
+    out = impl_ret_borrowed(context, builder, sig.return_type, res)
```
Comment (pitrou) [inline numba/targets/arrayobj.py]: Why is it borrowed? Is it the array object which is borrowed? As far as I can see the iterator object itself is new.

File: numba/runtime/_nrt_python.c (hunk 3 above)
Comment (sklam) [inline numba/runtime/_nrt_python.c]: this's mostly me guarding against bug during devel.  It also serves a documentation purpose.

File: numba/lowering.py (hunk 2 above)
Comment (sklam) [inline numba/lowering.py]: traces of my debugging

File: numba/targets/arrayobj.py (hunk 4 above)
Comment (sklam) [inline numba/targets/arrayobj.py]: I am taking advantage of the fact that ``context.nrt_incref`` traverse the structure to incref all ``MemInfo*`` (same for ``nrt_decref``).  I will change this to make it reflect what it is actually doing.  But, note that the decref will be applying on the ``iterobj``

**Pull request merged**